### PR TITLE
vk: Misc enhancements

### PR DESCRIPTION
**Pull request by kd-11**
1. Ports my old implementation of VK_EXT_device_fault to current rpcs3 version. This work was completed years ago but the branch sat around on local branch. Support for it is fairly widespread now, when the code was written there were 0 vendors shipping the extension. ~~Sadly, the most crashy driver for us (NVIDIA) does not support this feature.~~ (it is now supported by all vendors)
2. Add basic support for honeykrisp driver (asahi linux). The driver leaks memory like crazy in AAA titles and crashes but simpler titles run fine.

### CODE REVIEW

File: rpcs3/Emu/RSX/VK/VKHelpers.cpp [hunk 1]
```diff
@@ -139,6 +139,9 @@ namespace vk
 			rsx_log.error("Dozen is currently unsupported. How did you even get this to run outside windows?");
 #endif
 			break;
+		case driver_vendor::HONEYKRISP:
+			// Needs more testing
+			break;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/VKHelpers.cpp:144]: missing break below

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 2]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+		{
+			.sType = VK_STRUCTURE_TYPE_DEVICE_FAULT_INFO_EXT,
+			.pAddressInfos = address_info.data(),
+			.pVendorInfos = vendor_info.data()
+		};
+
+		fault_counts.vendorInfoCount = 0;
+		g_render_device->_vkGetDeviceFaultInfoEXT(*g_render_device, &fault_counts, &fault_info);
+
+		fault_message += fmt::format("Fault Summary:\n  %s\n\n", fault_info.description);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:38]:
```suggestion
		fmt::append(fault_message, "Fault Summary:\n  %s\n\n", fault_info.description);
```

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 3]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+
+		if (!address_info.empty())
+		{
+			fault_message += fmt::format("  Address Fault Information:\n", fault_info.description);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:42]:
```suggestion
			fmt::append(fault_message, "  Address Fault Information:\n", fault_info.description);
```

File: rpcs3/Emu/RSX/VK/VKHelpers.cpp (hunk 1 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKHelpers.cpp:144]: You mean the default tail block? That doesn't need a break statement.

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 4]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+					access_type = "instruction_pointer_unknown"; break;
+				case VK_DEVICE_FAULT_ADDRESS_TYPE_INSTRUCTION_POINTER_INVALID_EXT:
+					access_type = "instruction_pointer_invalid"; break;
+				case VK_DEVICE_FAULT_ADDRESS_TYPE_INSTRUCTION_POINTER_FAULT_EXT:
+					access_type = "instruction_pointer_fault"; break;
+				default:
+					break;
+				}
+
+				fault_message += fmt::format("  - Fault at address 0x%llx caused by %s\n", fault.reportedAddress, access_type);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:67]:
```suggestion
				fmt::append(fault_message, "  - Fault at address 0x%llx caused by %s\n", fault.reportedAddress, access_type);
```

File: rpcs3/Emu/RSX/VK/VKHelpers.cpp (hunk 1 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/VK/VKHelpers.cpp:144]: That's what we always do...
Weird reason to break code style and best practises now

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 5]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+			}
+		}
+
+		if (!vendor_info.empty())
+		{
+			fault_message += fmt::format("  Vendor Fault Information:\n", fault_info.description);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:73]:
```suggestion
			fmt::append(fault_message, "  Vendor Fault Information:\n", fault_info.description);
```

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 6]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+
+			for (const auto& fault : vendor_info)
+			{
+				fault_message += fmt::format("  - [0x%llx, 0x%llx] %s\n", fault.vendorFaultCode, fault.vendorFaultData, fault.description);
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:77]:
```suggestion
				fmt::append(fault_message, "  - [0x%llx, 0x%llx] %s\n", fault.vendorFaultCode, fault.vendorFaultData, fault.description);
```

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 7]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+		{
+			fault_message += fmt::format("  Address Fault Information:\n", fault_info.description);
+
+			for (const auto& fault : address_info)
+			{
+				std::string access_type = "unknown";
+				switch (fault.addressType)
+				{
+				case VK_DEVICE_FAULT_ADDRESS_TYPE_NONE_EXT:
+					access_type = "none"; break;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:50]: Also, strange code style here.
But that's just a nitpick

File: rpcs3/Emu/RSX/VK/vkutils/shared.cpp [hunk 8]
```diff
@@ -9,9 +10,80 @@ namespace vk
...
+
+		fault_message += fmt::format("Fault Summary:\n  %s\n\n", fault_info.description);
+
+		if (!address_info.empty())
+		{
+			fault_message += fmt::format("  Address Fault Information:\n", fault_info.description);
+
+			for (const auto& fault : address_info)
+			{
+				std::string access_type = "unknown";
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/VK/vkutils/shared.cpp:46]: you can move the "unknown" to the default case

File: rpcs3/Emu/RSX/VK/VKHelpers.cpp (hunk 1 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/VK/VKHelpers.cpp:144]: This isn't new code.

**Comment by kd-11** [comment]
Drafting while investigating https://github.com/RPCS3/rpcs3/issues/15289

**Pull request merged**